### PR TITLE
Update cyclic dependancy error message to show the chain of the cycle

### PR DESCRIPTION
**Pull request by Bryan-Cee**
This PR updates the cyclic dependency error message to show the nodes involved in the cyclical chain.

**Before:**

Screenshot 2021-10-12 at 11 16 47

**After:**

Screenshot 2021-10-12 at 12 16 56

### CODE REVIEW

File: src/graphHasCycles.ts [hunk 1]
```diff
@@ -90,9 +85,9 @@ const hasCycleDFS = (graph: Map<string, PGraphNodeWithDependencies>, visitMap: M
       }
 
       /**
-       * Add the current node's dependencies to the stack
+       * Add the current node's dependents to the stack
        */
-      stack.push(...[...node.dependsOn].map((n) => ({ node: n, traversing: false })));
+      stack.push(...[...node.dependedOnBy].map((n) => ({ node: n, traversing: false })));
```
Comment (VincentBailly) [inline src/graphHasCycles.ts:90]: Why do we need to spread node.dependedOnBy?

File: src/graphHasCycles.ts [hunk 2]
```diff
@@ -13,23 +13,17 @@ export function graphHasCycles(pGraphDependencyMap: Map<string, PGraphNodeWithDe
...
-      if (hasCycleDFS(pGraphDependencyMap,  visitMap, nodeId)) {
-        return {
-          hasCycle: true,
-          details: {
-            nodeId,
-            dependsOn: Array.from(nodes.dependsOn),
-            dependedOnBy: Array.from(nodes.dependedOnBy)
-          }
-        };
+      const cycle = hasCycleDFS(pGraphDependencyMap,  visitMap, nodeId);
```
Comment (VincentBailly) [inline src/graphHasCycles.ts]: the name of the function should be solved, the current name implies that the function returns a boolean

File: src/__tests__/index.tests.ts [hunk 3]
```diff
@@ -202,7 +202,10 @@ describe("Public API", () => {
       ["C", "D"],
       ["D", "B"],
     ];
-    const expectedErrorMessage = "The dependency graph has a cycle at B which depends on A,D and is depended on by C";
+    const expectedErrorMessage = `A cycle has been detected including the following nodes:
```
Comment (VincentBailly) [inline src/__tests__/index.tests.ts:205]: more test cases are needed: what happens if the graph has more than one cycle? what happens if the graph has two cycles which overlap?

Comment (Bryan-Cee) [reply src/__tests__/index.tests.ts:205]: Interesting view, let me update the tests with different scenarios to capture other edge cases.

File: src/graphHasCycles.ts (hunk 1 above)
Comment (Bryan-Cee) [reply src/graphHasCycles.ts:90]: The change does not change how the function works but changes the order of how we traverse the nodes. This results in the stack having the nodes from a parent-to-child order which makes it easier to list out the node where the cyclical dependency will have been found. For example, if we have a cycle where `A <= B <= C <= D <= B`, the `<=` means depended on by - where B depends on A and so on... 
When we spread `node.dependsOn` the `listOfCycle` in line 62 ends up having the nodes in the following order `B <= D <= C` when a cycle should have output `B <= C <= D`

Comment (VincentBailly) [reply src/graphHasCycles.ts:90]: Cool

File: src/__tests__/index.tests.ts [hunk 4]
```diff
@@ -209,6 +209,29 @@ D`;
...
+    const dependencies: DependencyList = [
+      ["A", "B"],
+      ["B", "C"],
+      ["C", "D"],
+      ["D", "B"],
+      ["C", "E"],
+      ["E", "F"],
+      ["F", "D"],
+    ];
+    expect(() => pGraph(nodeMap, dependencies)).toThrow();
```
Comment (VincentBailly) [inline src/__tests__/index.tests.ts]: What is the output in this case?

Comment (Bryan-Cee) [reply src/__tests__/index.tests.ts]: `B <- C <- E <- F <- D` is the first cycle detected

Comment (Bryan-Cee) [reply src/__tests__/index.tests.ts]: I have updated the test to depict the expected error message.
